2018/003-offline-chart-repository.md

```diff
@@ -0,0 +1,441 @@
+individual chart packages into the final image.
+
+The usage of the RPM patterns allows the image definition to stay lean and
+greatly reduces the amount of changes to perform against it.
+
+Finally, by using RPMs to ship the charts, we will be 100% sure multiple teams are not going to provide conflicting `.tgz` files (as in the same helm chart being provided by two different teams). Such a situation will be immediately result in a package conflict and prevent the final container image from being created.
+
+#### Maintenance operations
+
+This is an overview of the possible operations related with the helm charts and how these operations would translate against the open build service.
+
+  * Adding new helm charts: a new RPM for the chart is created. One of the RPM
+    pattern is updated to reference this new chart package. No change is needed
+    inside of the container image definition; the build service will automatically
+    rebuild it.
+  * Upgrading an existing helm chart: the RPM defining the helm chart is updated,
+    no change is needed inside of the RPM pattern requiring it, no change is
+    required inside of the container image definition. OBS will automatically
+    rebuild the image. Note well: adding a new version of an helm chart
+            falls into this category.
+  * Removing a chart: the RPM defining the chart has to be removed from the
+    project. The RPM pattern referencing the chart is updated. No change
+    is needed inside of the container image definition. OBS will automatically
+    rebuild the image.
+  * Add a new helm chart pattern: a new RPM defining the pattern is added to
+    the project, probably together with a bunch of new packages providing the
+    charts referenced inside of it. The container image definition is updated
+    to require this new pattern. OBS will automatically rebuild the image.
+  * Remove a helm chart pattern: the RPM pattern is removed from the project
+    together with all the packages that were being referenced by it. The
+    container image is going to be updated to remove the reference of the
+    pattern. OBS will automatically rebuild the container image.
+
+In each case it will be necessary to perform a OBS submission request from the
+devel project against the target project used by Maintenance QA. From this final
+project the `suse/kubernetes-charts` image is going to be published on the SUSE
+container registry.
+
+If the helm chart is referring a new set of container images, the submission
+request can include also the new container images. By doing that we ensure the
+helm charts and the contents of the SUSE registry are always aligned; we do not
+risk of having a helm chart referencing a container image that has not yet been
+published on the SUSE registry.
+**TODO:** is this possible? I mean having a SR that includes packages from one
+project (the devel prj of charts) and another one (the prj where the image
+referenced by the new chart is built).
+
```

Comment:
SRs/MRs have one source project. However, multiple SR/MR can be merged into the same incident/staging project. This requires coordination between the maintainers of the image and the maintainers of the packages. Actually in all cases the image needs to be submitted, so the coordination is always needed.

When you add a new helm chart pattern, the image won't build unless this packages is available. Thus packages should be "visible" from the project where we build the image since the beginning.

I think packages containing helm charts should live in the same project as the image which will contain them. This can be a bottleneck, I know, but other solutions like using links or aggregates or adding extra paths to the project, can be a mess.

If we did that, having the packages and the images in the same repo, then submitting them all together would be easy.

Reply:
+1 on storing the helm chart with the project (Dockerfile) on github. Unless you meant OBS?